## Handwriting Recognition using Transfer Learning

### Important imports

In [ ]:
import numpy as np
import os
import tensorflow as tf
import tensorflow_datasets as tfds
import pathlib
import matplotlib.pyplot as plt

Unzip the loaded files

In [ ]:
!unzip /content/ML1.zip
!unzip /content/ML2.zip

Initialise the training input and output(Hard coded)

In [ ]:
x_train = []
y_train = [[1,0,0,0,0,0],
           [1,0,0,0,0,0],
           [1,0,0,0,0,0],
           [1,0,0,0,0,0],
           
           [0,1,0,0,0,0],
           [0,1,0,0,0,0],
           [0,1,0,0,0,0],
           [0,1,0,0,0,0],
           
           [0,0,1,0,0,0],
           [0,0,1,0,0,0],
           [0,0,1,0,0,0],
           [0,0,1,0,0,0],
           
           [0,0,0,1,0,0],
           [0,0,0,1,0,0],
           [0,0,0,1,0,0],
           [0,0,0,1,0,0],
           
           [0,0,0,0,1,0],
           [0,0,0,0,1,0],
           [0,0,0,0,1,0],
           [0,0,0,0,1,0],
           
           [0,0,0,0,0,1],
           [0,0,0,0,0,1],
           [0,0,0,0,0,1],
           [0,0,0,0,0,1],
           
           [0,0,0,0,1,0],
           [0,0,0,0,0,1],
           [0,1,0,0,0,0],
           [1,0,0,0,0,0],
           [0,0,1,0,0,0],
           [0,0,0,1,0,0],]

Read images to covert it to 3d array or array of 2d images in grayscale.

[Help taken from](https://stackoverflow.com/a/49070833)

In [ ]:
# Source code is taken and edited from [Stack Overflow](https://stackoverflow.com/a/49070833)

from PIL import Image
import sys
import os
import csv

#Useful function
def createFileList(myDir, format='.jpg'):
  fileList = []
  print(myDir)
  for root, dirs, files in os.walk(myDir, topdown=False):
      for name in files:
          if name.endswith(format):
              fullName = os.path.join(root, name)
              fileList.append(fullName)
  return fileList

# Setup image dataset for test
myFileList = createFileList('/content/ML1/')

for file in myFileList:
    print(file)
    img_file = Image.open(file)
    # img_file.show()

    # get original image parameters...
    width, height = img_file.size
    format = img_file.format
    mode = img_file.mode

    # Make image Greyscale
    img_grey = img_file.convert('L')
    #img_grey.save('result.png')
    #img_grey.show()

    # Save Greyscale values
    value = np.asarray(img_grey.getdata(), dtype=np.int).reshape((img_grey.size[1], img_grey.size[0]))
    value = value/255.0
    #value = value.flatten()
    #print(value)
    #with open("img_pixels.csv", 'a') as f:
        #writer = csv.writer(f)
        #writer.writerow(value)
    x_train.append(value)

# Setup data for train
myFileList = createFileList('/content/ML2/')

for f in myFileList:
    print("File : ",f)
    img_file = Image.open(f)
    # img_file.show()

    # get original image parameters...
    width, height = img_file.size
    format = img_file.format
    mode = img_file.mode

    # Make image Greyscale
    img_grey = img_file.convert('L')
    #img_grey.save('result.png')
    #img_grey.show()

    # Save Greyscale values
    value = np.asarray(img_grey.getdata(), dtype=np.int).reshape((img_grey.size[1], img_grey.size[0]))
    value = value/255.0
    #value = value.flatten()
    # print(value)
    #with open("img_pixels.csv", 'a') as f:
        #writer = csv.writer(f)
        #writer.writerow(value)

    x_train.append(value)

x_train

/content/ML1/
/content/ML1/Triangle/image3.jpg
/content/ML1/Triangle/image5.jpg
/content/ML1/Triangle/image4.jpg
/content/ML1/Triangle/image6.jpg
/content/ML1/Diamond/image21.jpg
/content/ML1/Diamond/image20.jpg
/content/ML1/Diamond/image22.jpg
/content/ML1/Diamond/image23.jpg
/content/ML1/Square/image17.jpg
/content/ML1/Square/image18.jpg
/content/ML1/Square/image16.jpg
/content/ML1/Square/image19.jpg
/content/ML1/Grid_4x4/image26.jpg
/content/ML1/Grid_4x4/image27.jpg
/content/ML1/Grid_4x4/image24.jpg
/content/ML1/Grid_4x4/image25.jpg
/content/ML1/Infinity/image15.jpg
/content/ML1/Infinity/image13.jpg
/content/ML1/Infinity/image12.jpg
/content/ML1/Infinity/image14.jpg
/content/ML1/Circle/image8.jpg
/content/ML1/Circle/image7.jpg
/content/ML1/Circle/image10.jpg
/content/ML1/Circle/image9.jpg
/content/ML2/
File :  /content/ML2/i.jpg
File :  /content/ML2/c.jpg
File :  /content/ML2/d.jpg
File :  /content/ML2/t.jpg
File :  /content/ML2/s.jpg
File :  /content/ML2/i (1).jpg


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:39: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:67: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecat

[array([[1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 0.99215686, 1.        , 1.        ,
         1.        , 0.99215686, 1.        , 1.        , 1.        ,
         1.        , 1.        , 1.        ],
        [1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 1.        , 0.98823529, 1.        ,
         1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 0.99607843, 0.99215686, 0.99607843,
         1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 1.        ],
        [1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 1.        , 1.        , 1.        , 1.        ,
         1.        , 0.98039216, 0.74509804, 0.83921569, 1.        ,
         1.

## Define and compile the first model

In [ ]:
from keras.backend import batch_normalization
import keras
from keras.models import Sequential
from keras.layers import Conv2D, Lambda, MaxPooling2D
from keras.layers import Dense, Dropout, Flatten
from  keras.layers.normalization.batch_normalization import BatchNormalization

from keras.preprocessing.image import ImageDataGenerator

from keras.utils.np_utils import to_categorical


model=Sequential()
   
model.add(Conv2D(filters=64, kernel_size = (3,3), activation="relu", input_shape=(28,28,1)))
model.add(Conv2D(filters=64, kernel_size = (3,3), activation="relu"))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(BatchNormalization())

model.add(Conv2D(filters=128, kernel_size = (3,3), activation="relu"))
model.add(Conv2D(filters=128, kernel_size = (3,3), activation="relu"))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(BatchNormalization())    

model.add(Conv2D(filters=256, kernel_size = (3,3), activation="relu"))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(BatchNormalization())
    
model.add(Flatten())
model.add(Dense(512,activation="relu"))
    
model.add(Dense(6,activation="softmax"))
    
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])


## Train the first model

In [ ]:
x_train = np.asarray(x_train)
y_train = np.asarray(y_train)
model.fit(x_train, y_train, batch_size=32, epochs=10, validation_split=0.2)

Epoch 1/10
1/1 [==============================] - 1s 1s/step - loss: 2.1018 - accuracy: 0.1250 - val_loss: 1.7899 - val_accuracy: 0.1667
Epoch 2/10
1/1 [==============================] - 0s 229ms/step - loss: 0.1642 - accuracy: 1.0000 - val_loss: 1.7859 - val_accuracy: 0.5000
Epoch 3/10
1/1 [==============================] - 0s 230ms/step - loss: 0.0269 - accuracy: 1.0000 - val_loss: 1.7813 - val_accuracy: 0.1667
Epoch 4/10
1/1 [==============================] - 0s 234ms/step - loss: 0.0120 - accuracy: 1.0000 - val_loss: 1.7777 - val_accuracy: 0.1667
Epoch 5/10
1/1 [==============================] - 0s 234ms/step - loss: 0.0076 - accuracy: 1.0000 - val_loss: 1.7740 - val_accuracy: 0.3333
Epoch 6/10
1/1 [==============================] - 0s 241ms/step - loss: 0.0053 - accuracy: 1.0000 - val_loss: 1.7710 - val_accuracy: 0.3333
Epoch 7/10
1/1 [==============================] - 0s 233ms/step - loss: 0.0037 - accuracy: 1.0000 - val_loss: 1.7690 - val_accuracy: 0.3333
Epoch 8/10
1/1 [=======

## Setup for the second model to transfer knowledge
Uses MNISt dataset to learn handwrtten characters.

In [ ]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

Remove the last layer from model and a new layer of size 10 to classify digits.

In [ ]:
model.pop()
model.add(Dense(10,activation="softmax"))

Freeze the layers except last and compile.

In [ ]:
for layer in model.layers[:-1]:
  layer.trainable = False


model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True), optimizer="adam", metrics=["accuracy"])

Get the summary of model.

In [ ]:
model.summary()

Model: "sequential_14"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_47 (Conv2D)          (None, 26, 26, 64)        640       
                                                                 
 conv2d_48 (Conv2D)          (None, 24, 24, 64)        36928     
                                                                 
 max_pooling2d_28 (MaxPoolin  (None, 12, 12, 64)       0         
 g2D)                                                            
                                                                 
 batch_normalization_26 (Bat  (None, 12, 12, 64)       256       
 chNormalization)                                                
                                                                 
 conv2d_49 (Conv2D)          (None, 10, 10, 128)       73856     
                                                                 
 conv2d_50 (Conv2D)          (None, 8, 8, 128)       

Train the new model.

In [ ]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5


/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/dispatch.py:1082: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a sigmoid or softmax activation and thus does not represent logits. Was this intended?"
  return dispatch_target(*args, **kwargs)


1875/1875 [==============================] - 107s 57ms/step - loss: 2.0175 - accuracy: 0.5398
Epoch 2/5
1875/1875 [==============================] - 107s 57ms/step - loss: 1.6040 - accuracy: 0.7224
Epoch 3/5
1875/1875 [==============================] - 107s 57ms/step - loss: 1.3388 - accuracy: 0.7668
Epoch 4/5
1875/1875 [==============================] - 107s 57ms/step - loss: 1.1539 - accuracy: 0.7912
Epoch 5/5
1875/1875 [==============================] - 108s 58ms/step - loss: 1.0175 - accuracy: 0.8094


Evalute the model.

In [ ]:
model.evaluate(x_test,  y_test, verbose=2)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/dispatch.py:1082: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a sigmoid or softmax activation and thus does not represent logits. Was this intended?"
  return dispatch_target(*args, **kwargs)


313/313 - 16s - loss: 0.9314 - accuracy: 0.8284 - 16s/epoch - 52ms/step


[0.9313758015632629, 0.8284000158309937]

Our new model gives accuracy of 82.84% on MNIST dataset.